src/Joblify.Core/User/UserService.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System;
+using System.Linq;
+using System.Threading.Tasks;
+using AutoMapper;
+using Joblify.Core.Data.UnitOfWork;
+using Joblify.Core.User.Dto;
+using Microsoft.EntityFrameworkCore;
+
+namespace Joblify.Core.User
+{
+    public class UserService : IUserService
+    {
+        private readonly IUnitOfWork _unitOfWork;
+        private readonly IMapper _mapper;
+
+        public UserService(IUnitOfWork unitOfWork, IMapper mapper)
+        {
+            _unitOfWork = unitOfWork;
+            _mapper = mapper;
+        }
+
+        public async Task<AddUserDto> CreateUser(AddUserDto userDto)
+        {
+            var role = await _unitOfWork.RoleRepository.Entities.SingleAsync(r => r.Name == userDto.RoleName);
+            if (role == null)
+            {
+                return null;
+            }
+
+            var externalProvider = await _unitOfWork.ExternalProviderRepository.Entities.SingleAsync(e => e.Name == userDto.ExternalProviderName);
+            if (externalProvider == null)
+            {
+                return null;
+            }
+
+            var userEntity = _mapper.Map<AddUserDto, Data.Models.User>(userDto);
+            userEntity.Role = role;
+            userEntity.ExternalProvider = externalProvider;
+
+            await _unitOfWork.UserRepository.AddAsync(userEntity);
+            await _unitOfWork.CommitAsync();
+
+            return _mapper.Map<Data.Models.User, AddUserDto>(userEntity);
+        }
+
+        public async Task<AddUserDto> UpdateUser(UpdateUserDto userDto)
+        {
+            var userEntity = _unitOfWork.UserRepository.Entities.SingleOrDefault(u => u.Email == userDto.Email);
+            if (userEntity == null)
+            {
+                throw new Exception("User does not exist");
```

Comment:
why throw?

Reply:
Agree. In fact I handle in controler posibility of getting null from repository. So it isn't redundancy here. However if I need to do this here, should I use try catch block with NullReferenceException?

Reply:
First of all you should think if you really need to throw something - here we have just null being returned, so it's perfect candidate for either 400 or 404 response.
Also, to address your question - yes, if you want to react to specific exceptions, you have to catch them and handle application flow in catch or finally (or middleware designated to handle server-side exceptions). Otherwise you don't have any control over what happens after exception being throw and app may simply "explode".

Reply:
Moreover I'm losing the original stack trace of the exception, as well as its type here. I fixed it in a commit below. Thx for review.